### PR TITLE
add cub scan back, pose bug fix and missing variations test added

### PR DESCRIPTION
**Pull request by JerryJiehanWang**

### CODE REVIEW

File: kaolin/csrc/cuda/ops/conversions/unbatched_mcube/unbatched_mcube_cuda_kernel.cu [hunk 1]
```diff
@@ -491,28 +495,52 @@ generateTriangles2(float *pos, int *faces, int *compactedVoxelArray,
...
-                                          maxVerts);
+    }
+
+    generateTriangles2<<<grid2, NTHREADS>>>(pos.data_ptr<float>(), faces.data_ptr<int>(),
+                                            compactedVoxelArray.data_ptr<int>(), numTrianglesScanned.data_ptr<int>(),
+                                            numPartialVertsScanned.data_ptr<int>(), numPartialVerts.data_ptr<int>(),
+                                            voxelVertsOrder.data_ptr<int>(),
+                                            voxelgrid.data_ptr<float>(), gridSize,
+                                            voxelSize, isoValue, activeVoxels,
+                                            maxVerts);
```
Comment (Caenorst) [inline kaolin/csrc/cuda/ops/conversions/unbatched_mcube/unbatched_mcube_cuda_kernel.cu]: Not sure why the spacing changed (auto-indentation with your IDE ?) with C++ we usually follow [the Google style guide](https://google.github.io/styleguide/cppguide.html) , which means 2 space for this kind of indentation. Although we've losely enforced it, I prefer avoid any modification that would go against it and make the review harder.

File: kaolin/csrc/cuda/ops/conversions/unbatched_mcube/unbatched_mcube_cuda_kernel.cu [hunk 2]
```diff
@@ -491,28 +495,52 @@ generateTriangles2(float *pos, int *faces, int *compactedVoxelArray,
...
+    int *d_in = input.data_ptr<int>();
+    int *d_out = output.data_ptr<int>();
+
+    cudaError_t err = cudaGetLastError();
+
+    if ( err != cudaSuccess )
+    {
+    printf("CUDA Error: %s\n", cudaGetErrorString(err));
+    exit(-1);
+    }
```
Comment (Caenorst) [inline kaolin/csrc/cuda/ops/conversions/unbatched_mcube/unbatched_mcube_cuda_kernel.cu]: What is the reason to put an error catching **before** executing the function ?

File: kaolin/csrc/cuda/ops/conversions/unbatched_mcube/unbatched_mcube_cuda_kernel.cu [hunk 3]
```diff
@@ -491,28 +495,52 @@ generateTriangles2(float *pos, int *faces, int *compactedVoxelArray,
...
+    // Determine temporary device storage requirements
+    void *d_temp_storage = NULL;
+    size_t temp_storage_bytes = 0;
+    kaolin::cub::DeviceScan::ExclusiveSum(d_temp_storage, temp_storage_bytes, d_in, d_out, numElements);
+
+    // Allocate temporary storage
+    cudaMalloc(&d_temp_storage, temp_storage_bytes);
```
Comment (Caenorst) [inline kaolin/csrc/cuda/ops/conversions/unbatched_mcube/unbatched_mcube_cuda_kernel.cu]: Avoid cudaMalloc, prefer torch::empty / zeros.

File: setup.py [hunk 4]
```diff
@@ -55,8 +55,8 @@
 
 import numpy
 import torch
-from torch.utils.cpp_extension import BuildExtension, CppExtension, \
-    CUDAExtension
+from torch.utils.cpp_extension import BuildExtension, CppExtension, CUDAExtension
+from torch.utils.cpp_extension import CUDA_HOME as cuda_home
```
Comment (Caenorst) [inline setup.py]: `cuda_home` is unused, it was for debugging

File: tests/python/kaolin/ops/conversions/test_voxelgrid.py [hunk 5]
```diff
@@ -1074,50 +1074,61 @@ def _flip(self, voxelgrid, expected_vertices, dims):
...
+            new_expected_vertices, new_voxelgrid = self._rotate(voxelgrid, expected_vertices, rotate)
             for flip in all_flips:
-                new_vertices, new_expected_vertices, _ = self._flip(voxelgrid, expected_vertices, flip)
-                new_vertices = torch.sort(new_vertices[0], dim=0)[0]
-                new_expected_vertices = torch.sort(new_expected_vertices, dim=0)[0]
+                curr_expected_vertices, curr_voxelgrid = self._flip(new_voxelgrid, new_expected_vertices, flip)
+                curr_vertices, _ = vg.voxelgrids_to_trianglemeshes(curr_voxelgrid.unsqueeze(0))
+
+                print(curr_expected_vertices)
+                print(curr_vertices)
```
Comment (Caenorst) [inline tests/python/kaolin/ops/conversions/test_voxelgrid.py]: Can remove print.

File: tests/python/kaolin/ops/conversions/test_voxelgrid.py [hunk 6]
```diff
@@ -1074,50 +1074,61 @@ def _flip(self, voxelgrid, expected_vertices, dims):
...
 
         for rotate in all_rotations:
-            vertices, expected_vertices, voxelgrid = self._rotate(voxelgrid, expected_vertices, rotate)
+            new_expected_vertices, new_voxelgrid = self._rotate(voxelgrid, expected_vertices, rotate)
             for flip in all_flips:
-                new_vertices, new_expected_vertices, _ = self._flip(voxelgrid, expected_vertices, flip)
-                new_vertices = torch.sort(new_vertices[0], dim=0)[0]
-                new_expected_vertices = torch.sort(new_expected_vertices, dim=0)[0]
+                curr_expected_vertices, curr_voxelgrid = self._flip(new_voxelgrid, new_expected_vertices, flip)
+                curr_vertices, _ = vg.voxelgrids_to_trianglemeshes(curr_voxelgrid.unsqueeze(0))
```
Comment (Caenorst) [inline tests/python/kaolin/ops/conversions/test_voxelgrid.py:1123]: Good to have the `voxelgrids_to_trianglemeshes` here instead of _rotate / _flip.

**Review by Caenorst — COMMENTED** [review]
Looking good, just the indentation which seems automatically modified, and the cudaMalloc to remove.

File: kaolin/csrc/cuda/ops/conversions/unbatched_mcube/unbatched_mcube_cuda_kernel.cu (hunk 1 above)
Comment (JerryJiehanWang) [reply kaolin/csrc/cuda/ops/conversions/unbatched_mcube/unbatched_mcube_cuda_kernel.cu]: Maybe it's a copy-paste issue, fixed.

File: kaolin/csrc/cuda/ops/conversions/unbatched_mcube/unbatched_mcube_cuda_kernel.cu (hunk 2 above)
Comment (JerryJiehanWang) [reply kaolin/csrc/cuda/ops/conversions/unbatched_mcube/unbatched_mcube_cuda_kernel.cu]: For debugging, fixed now.

File: setup.py (hunk 4 above)
Comment (JerryJiehanWang) [reply setup.py]: removed.

File: tests/python/kaolin/ops/conversions/test_voxelgrid.py (hunk 5 above)
Comment (JerryJiehanWang) [reply tests/python/kaolin/ops/conversions/test_voxelgrid.py]: removed.

File: kaolin/csrc/cuda/ops/conversions/unbatched_mcube/unbatched_mcube_cuda_kernel.cu (hunk 3 above)
Comment (JerryJiehanWang) [reply kaolin/csrc/cuda/ops/conversions/unbatched_mcube/unbatched_mcube_cuda_kernel.cu]: Not exactly sure how to use torch::empty/zeros here: https://nvlabs.github.io/cub/structcub_1_1_device_scan.html#a02b2d2e98f89f80813460f6a6ea1692b

Comment (JerryJiehanWang) [reply kaolin/csrc/cuda/ops/conversions/unbatched_mcube/unbatched_mcube_cuda_kernel.cu]: fixed

**Review by Caenorst — DISMISSED** [review]
LGTM! Thanks!

**Comment by Caenorst** [comment]
It seems that submodules are not loaded in the github ci part: can you follow recommendations from https://docs.drone.io/pipeline/docker/syntax/cloning/ (see the section "The --recursive flag")

**Review by Caenorst — APPROVED** [review]
Good job ! Thanks for the fix ! :)

**Pull request merged**